### PR TITLE
feat: (sqlite) support namespace at storage level, mostly for flags

### PR DESCRIPTION
**Pull request by markphelps**
feat: (sqlite) wip support namespace at storage level, mostly for flags

fixes: FLI-237

basic storage code for adding namespaces for flags/variants. ONLY works with SQLite, other dbs will come once SQLite is fully implemented to prevent wasted effort

### CODE REVIEW

File: internal/server/middleware/grpc/middleware.go [hunk 1]
```diff
@@ -234,16 +234,29 @@ func CacheUnaryInterceptor(cache cache.Cacher, logger *zap.Logger) grpc.UnarySer
...
-func flagCacheKey(key string) string {
-	k := fmt.Sprintf("f:%s", key)
+func flagCacheKey(namespaceKey, key string) string {
+	var k string
+	// for backwards compatibility
+	if namespaceKey != "" {
+		k = fmt.Sprintf("f:%s:%s", namespaceKey, key)
+	} else {
+		k = fmt.Sprintf("f:%s", key)
+	}
```
Comment (yquansah) [inline internal/server/middleware/grpc/middleware.go:258]: The backwards compatibility here refers to other dbs as a data source other than sqlite correct? Or is there more it is referring to?

File: config/migrations/sqlite3/6_create_namespaces.up.sql [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  FOREIGN KEY (namespace_key, flag_key) REFERENCES flags (namespace_key, key) ON DELETE CASCADE
+);
+
+INSERT INTO rules_temp (id, flag_key, segment_key, rank, created_at, updated_at)
+  SELECT id, flag_key, segment_key, rank, created_at, updated_at
+  FROM rules;
+
+DROP TABLE rules;
+
+ALTER TABLE rules_temp RENAME TO rules;
```
Comment (GeorgeMac) [inline config/migrations/sqlite3/6_create_namespaces.up.sql:83]: Could we forgo namespacing both rules and variants?

My thinking is that they're already "namespaced" beneath a flag.
Their API already appears to require a flag key on which to group operations.
So by extension, they're nested beneath the namespace in the relational hierarchy.

File: internal/storage/sql/common/flag.go [hunk 3]
```diff
@@ -54,7 +59,7 @@ func (s *Store) GetFlag(ctx context.Context, key string) (*flipt.Flag, error) {
 
 	if err != nil {
 		if errors.Is(err, sql.ErrNoRows) {
-			return nil, errs.ErrNotFoundf("flag %q", key)
+			return nil, errs.ErrNotFoundf("flag %q; namespace %q", key, namespaceKey)
```
Comment (GeorgeMac) [inline internal/storage/sql/common/flag.go:62]: Pedantic mode: How about switching the order? Just to preserve left-to-right precedence.
```suggestion
			return nil, errs.ErrNotFoundf("namespace %q; flag %q", namespaceKey, key)
```
or even something like the following:
```suggestion
			return nil, errs.ErrNotFoundf(`flag "%s/%s"`, namespaceKey, key)
```

File: internal/storage/sql/common/flag.go [hunk 4]
```diff
@@ -281,10 +298,14 @@ func (s *Store) ListFlags(ctx context.Context, opts ...storage.QueryOption) (sto
...
-func (s *Store) CountFlags(ctx context.Context) (uint64, error) {
+func (s *Store) CountFlags(ctx context.Context, namespaceKey string) (uint64, error) {
 	var count uint64
 
-	if err := s.builder.Select("COUNT(*)").From("flags").QueryRowContext(ctx).Scan(&count); err != nil {
+	if namespaceKey == "" {
+		namespaceKey = storage.DefaultNamespace
+	}
+
+	if err := s.builder.Select("COUNT(*)").From("flags").Where(sq.Eq{"namespace_key": namespaceKey}).QueryRowContext(ctx).Scan(&count); err != nil {
```
Comment (GeorgeMac) [inline internal/storage/sql/common/flag.go:308]: Take it or leave it: Maybe break this across multiple lines?

File: config/migrations/sqlite3/6_create_namespaces.up.sql [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+  namespace_key VARCHAR(255) NOT NULL DEFAULT 'default' REFERENCES namespaces ON DELETE CASCADE,
+  PRIMARY KEY (namespace_key, key)); /* composite primary key unique by default */
+
+INSERT INTO flags_temp (key, name, description, enabled, created_at, updated_at)
+  SELECT key, name, description, enabled, created_at, updated_at
+  FROM flags;
+
+DROP TABLE flags;
+
+ALTER TABLE flags_temp RENAME TO flags;
```
Comment (GeorgeMac) [inline config/migrations/sqlite3/6_create_namespaces.up.sql:40]: Genuine open question: Is there any value in breaking the creation of the namespaces table and the migration of the flags table into separate migrations? 🤔 

Maybe only for clarity. I doubt there is little functional value to us, but food for thought.

**Review by GeorgeMac — CHANGES_REQUESTED** [review]
The change looks good, just had a couple style suggestions and mainly one open question.

Is it possible that we could forgo adding namespaces to things already scoped to other resources?

Just thinking outloud. It is a little disconcerting that we have to lift and shift a whole table.
But I suspect there is no way around it. Even if we went v2 we would have to help folks migrate and do something similar. This is definitely something to keep pondering on (the migration process to a namespaced future).

File: config/migrations/sqlite3/6_create_namespaces.up.sql (hunk 2 above)
Comment (GeorgeMac) [reply config/migrations/sqlite3/6_create_namespaces.up.sql]: Makes total sense.

File: internal/storage/sql/common/flag.go [hunk 6]
```diff
@@ -34,16 +34,21 @@ func emptyAsNil(str string) *string {
...
 		createdAt fliptsql.Timestamp
 		updatedAt fliptsql.Timestamp
 
 		flag = &flipt.Flag{}
 
-		err = s.builder.Select("\"key\", name, description, enabled, created_at, updated_at").
+		err = s.builder.Select("namespace_key, \"key\", name, description, enabled, created_at, updated_at").
 			From("flags").
-			Where(sq.Eq{"\"key\"": key}).QueryRowContext(ctx).Scan(
+			Where(sq.And{sq.Eq{"namespace_key": namespaceKey}, sq.Eq{"\"key\"": key}}).QueryRowContext(ctx).Scan(
```
Comment (GeorgeMac) [inline internal/storage/sql/common/flag.go:50]: Take it or leave it: I think you can do this in a few places:
```suggestion
			Where(sq.Eq{"namespace_key": namespaceKey, \"key\"": key}).QueryRowContext(ctx).Scan(
```

**Review by GeorgeMac — APPROVED** [review]
The namespace tendrils are coming.
One take it or leave it. Plus I like your suggestion in the `TODO` in the tests to cover more than one namespace (more than just `default`). Otherwise, this change looks good to me.

**Review by yquansah — APPROVED** [review]
Code looks smooth.

**Pull request merged**